flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,377 @@
+					assertNull(decodedBufferResponse.getBuffer());
+				} else {
+					assertEquals(expectedBufferResponse.getBuffer(), decodedBufferResponse.getBuffer());
+				}
+			} else if (expected[i] instanceof NettyMessage.ErrorResponse) {
+				ErrorResponse expectedErrorResponse = (ErrorResponse) expected[i];
+				ErrorResponse decodedErrorResponse = (ErrorResponse) decodedMessages.get(i);
+
+				assertEquals(expectedErrorResponse.receiverId, decodedErrorResponse.receiverId);
+				assertEquals(expectedErrorResponse.cause.getClass(), decodedErrorResponse.cause.getClass());
+				assertEquals(expectedErrorResponse.cause.getMessage(), decodedErrorResponse.cause.getMessage());
+			} else {
+				fail("Unsupported message type");
+			}
+		}
+	}
+
+	private ByteBuf[] serializeMessages(NettyMessage[] messages) throws Exception {
+		ByteBuf[] serializedBuffers = new ByteBuf[messages.length];
+		for (int i = 0; i < messages.length; ++i) {
+			serializedBuffers[i] = messages[i].write(ALLOCATOR);
+		}
+
+		return serializedBuffers;
+	}
+
+	private int[] getBufferSizes(ByteBuf[] buffers) {
+		int[] sizes = new int[buffers.length];
+		for (int i = 0; i < sizes.length; ++i) {
+			sizes[i] = buffers[i].readableBytes();
+		}
+
+		return sizes;
+	}
+
+	private ByteBuf mergeBuffers(ByteBuf[] buffers) {
+		ByteBuf allData = ALLOCATOR.buffer();
+		for (ByteBuf buffer : buffers) {
+			allData.writeBytes(buffer);
+		}
+
+		return allData;
+	}
+
+	private ByteBuf[] partitionBuffer(ByteBuf buffer, int[] partitionPositions) {
+		ByteBuf[] result = new ByteBuf[partitionPositions.length + 1];
+		for (int i = 0; i < result.length; ++i) {
+			int startPos = (i == 0 ? 0 : partitionPositions[i - 1]);
+			int endPos = (i < partitionPositions.length ? partitionPositions[i] : buffer.readableBytes());
+
+			result[i] = ALLOCATOR.buffer();
+			result[i].writeBytes(buffer, startPos, endPos - startPos);
+		}
+
+		return result;
+	}
+
+	private void releaseBuffers(ByteBuf[] buffers) {
+		for (ByteBuf buffer : buffers) {
+			buffer.release();
+		}
+	}
+
+	private Buffer createDataBuffer(int size) {
+		MemorySegment segment = MemorySegmentFactory.allocateUnpooledSegment(size);
+		NetworkBuffer buffer = new NetworkBuffer(segment, FreeingBufferRecycler.INSTANCE);
+		for (int i = 0; i < size / 4; ++i) {
+			buffer.writeInt(i);
+		}
+
+		return buffer;
+	}
+
+	private CreditBasedPartitionRequestClientHandler createPartitionRequestClientHandler(int numberOfBuffersInNormalChannel) throws IOException {
+		NetworkBufferPool networkBufferPool = new NetworkBufferPool(
+			numberOfBuffersInNormalChannel,
+			32 * 1024,
+			numberOfBuffersInNormalChannel);
+
+		SingleInputGate singleInputGate = createSingleInputGate(2);
+
+		CreditBasedPartitionRequestClientHandler handler = new CreditBasedPartitionRequestClientHandler();
+
+		RemoteInputChannel normalInputChannel = spy(new RemoteInputChannel(
+			singleInputGate,
+			0,
+			new ResultPartitionID(),
+			mock(ConnectionID.class),
```

Comment:
we should avoid mocking if tests, also for other places.

Reply:
Replaced the `mock` method with `InputGateBuilder` and `InputChannelBuilder`.